File: Data Structures & Algorithms/linked-list-cycle-detection/submission-5.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @return {boolean}
     */
    hasCycle(head) {
        let slow = head, fast = head;
        if(slow === null) return false;
        while(fast.next !== null && fast.next.next !== null){
            fast = fast.next.next;
            slow = slow.next;
            if(fast === slow){
                return true;
            }
        }
        return false;
    }
}

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-5.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        minHeap = []
        count = 0

        for node in lists:
            if node:
                heapq.heappush(minHeap, (node.val, count, node))
                count += 1
        
        res = ListNode(0)
        r = res

        while minHeap:
            val, count, node = heapq.heappop(minHeap)

            res.next = node

            if node and node.next:
                heapq.heappush(minHeap, (node.next.val, count, node.next))
            count += 1
            res = res.next

        return r.next
       